homeassistant/components/solaredge_local/sensor.py

```diff
@@ -167,6 +252,14 @@ def unit_of_measurement(self):
         """Return the unit of measurement."""
         return self._unit_of_measurement
 
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        try:
+            return {self._attr: self._data.info[self._json_key]}
+        except KeyError:
```

Comment:
Is the KeyError and missing attribute key always correlated? Are we risking to return a dict with `None` as key?

Reply:
No i don't think so.
https://github.com/home-assistant/home-assistant/blob/58e5c553f2138a18d9d6b3f44040a9f98a94c5b5/homeassistant/components/solaredge_local/sensor.py#L346
In this line the data is created in the info dict with the name provided by: `self._json_key` in this case: `"optimizers"`
So if there is no data created in the info dict then it would return `KeyError` boucouse the name `self._json_key` does not exist.

Reply:
Right, but are we sure that this `KeyError` case correlates to when `self._attr` is `None`? I'm worried since we use two different variables `self._attr` and self._json_key` but we only check one of them here with the `KeyError` exception, `self._json_key`.

Reply:
This is what the states menu looks like:
![states menu](https://user-images.githubusercontent.com/38077357/66413245-4f693680-e9f7-11e9-962a-bbf081d8468a.PNG)

Reply:
Does HA only check the first one? If that is so then it should not work.
but if I try this in an IDE
```python
import traceback

test1 = None
test2 = {}
test2["temperature"] = [
        "currentTemp",
        "Current temperature",
        'C',
        "mdi:oil-temperature"]

try:
    print("beginning try")
    print(test1)

    print(test2['temperature'])
    print(test2['nottemperature'])
    #traceback.print_exc()
except:
    pass
    traceback.print_exc()
```
```txt

beginning try
None
['currentTemp', 'Current temperature', 'C', 'mdi:oil-temperature']
Traceback (most recent call last):
  File "D:/Users/eric/OneDrive - Hanze/OneDrive - Hanzehogeschool Groningen/programeren/random test/append en add dicts test.py", line 108, in <module>
    print(test2['nottemperature'])
KeyError: 'nottemperature'

```

So if HA treats the `try` the same as my IDE it should work because printing `test1` that has value ` None` does not create a problem.

Reply:
The `KeyError` happens if `self._json_key` is not in `self._data.info` dict. It doesn't take the `self._attr` value into account at all. We have to make that connection separately or add another check.

Reply:
Currently the connection is done manually here:
```py
"optimizer_connected": [
        "optimizers",
        "Optimizers Online",
        "optimizers",
        "mdi:solar-panel",
        "optimizers_connected",
    ],
```
and here:
```py
self.info["optimizers"] = status.optimizersStatus.total
self.info["invertertemperature"] = INVERTER_MODES[status.status]
```
So if we would make an error in those two places and not match both places, we'll get a non expected result in the `device_state_attributes`. So if we can make this connection more strict, it would be less easy to make a mistake.

Reply:
Okay, I came to the conclusion that `self._attr` in `\/` does not create an error.   [picture below] https://github.com/home-assistant/home-assistant/blob/060ea848bc16173bddd8c76ee19bd5d8c5a94ca7/homeassistant/components/solaredge_local/sensor.py#L256-L262

It will always send Null to the attributes if a line like this exist.   [picture below]  https://github.com/home-assistant/home-assistant/blob/060ea848bc16173bddd8c76ee19bd5d8c5a94ca7/homeassistant/components/solaredge_local/sensor.py#L346-L347

So a problem occurs when a developer does not add an attribute name to the `SENSOR_TYPES` 
That error could be fixed with this:
```python
    @property
    def device_state_attributes(self):
        """Return the state attributes."""
        if self._attr:
            try:
                return {self._attr: self._data.info[self._json_key]}
            except KeyError:
                return None
```
With this code both places need to be occupied. So the sensor needs an attribute name and attribute data to be visible in the fronted.
Do you mean this check?


--------------------------
--------------------
--------------------


![null in sensor](https://user-images.githubusercontent.com/38077357/66422842-ba703880-ea0a-11e9-9551-f8dc7a9cd2ed.PNG)
This line was added: 
`self.info["optimizerpower"] = power`

Reply:
So a problem occurs when a developer does not add an attribute name to the SENSOR_TYPES
That error could be fixed with this:
```python
    @property
    def device_state_attributes(self):
        """Return the state attributes."""
        if self._attr:
            try:
                return {self._attr: self._data.info[self._json_key]}
            except KeyError:
                return None
```
With this code both places need to be occupied. So the sensor needs an attribute name and attribute data to be visible in the fronted.
**Do you mean this check?**
Correct me if I am wrong.

Reply:
Yes, that's enough. We should `return None` in the other case.